### PR TITLE
Skip certificate generation if SSL is disabled

### LINKED ISSUE #379: Skip SSL certificate generation
I use MySQL for local test, on each startup MySQL take some time to generate a random SSL certificate.

Is it possible to skip the SSL certificate generation?

### PR DESCRIPTION
**Pull request by tianon**
Closes #379

### CODE REVIEW

File: 8.0/docker-entrypoint.sh [hunk 1]
```diff
@@ -111,7 +111,8 @@ if [ "$1" = 'mysqld' -a -z "$wantHelp" ]; then
 		"$@" --initialize-insecure
 		echo 'Database initialized'
 
-		if command -v mysql_ssl_rsa_setup > /dev/null && [ ! -e "$DATADIR/server-key.pem" ]; then
+		sslEnabled="$(_get_config 'ssl' "$@")"
+		if [ "$sslEnabled" = 'TRUE' ] && command -v mysql_ssl_rsa_setup > /dev/null && [ ! -e "$DATADIR/server-key.pem" ]; then
```
[the remark this answers is not in the document]
Comment (tianon) [reply 8.0/docker-entrypoint.sh]: Friendly ping @ltangvald :heart:

**Comment by tianon** [comment]
https://github.com/docker-library/mysql/issues/379#issuecomment-420773317

> If the user supplies their own certificates in the appropriate place, MySQL should transparently pick them up and should not create new ones. If that behavior is not working, then we should look at fixing that separately.

File: 8.0/docker-entrypoint.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (tianon) [reply 8.0/docker-entrypoint.sh]: Friendly ping @ltangvald :wink:

> I was testing this and found that in 8.0, it appears that `--initalize-insecure` is creating a certificate itself if SSL is enabled -- should we remove this bit of code entirely from 8.0? (or are there cases where SSL is enabled and `--initalize-insecure` doesn't create a `server-key.pem` file?)

**Comment by tianon** [comment]
Rebased/updated to the new `docker-entrypoint.sh` structure

**Comment by ltangvald** [comment]
I _think_ the separate mysql_ssl_rsa_setup utility was mostly just relevant before we  started linking openssl (for licensing issues), so this may no longer be needed for any versions? Also, this behavior can be controlled with the auto_generate_certs and  sha256_password_auto_generate_rsa_keys settings. So by virtue of me being over 2 years late *cough*, I think this probably isn't needed any more :)

**Comment by tianon** [comment]
So if I'm understanding correctly, this change should be updated to the following, even on MySQL 5.6?

```diff
diff --git a/.template.Debian/docker-entrypoint.sh b/.template.Debian/docker-entrypoint.sh
index 6845ee4..6024a90 100755
--- a/.template.Debian/docker-entrypoint.sh
+++ b/.template.Debian/docker-entrypoint.sh
@@ -163,13 +163,6 @@ docker_init_database_dir() {
 		"$@" --initialize-insecure
 	fi
 	mysql_note "Database files initialized"
-
-	if command -v mysql_ssl_rsa_setup > /dev/null && [ ! -e "$DATADIR/server-key.pem" ]; then
-		# https://github.com/mysql/mysql-server/blob/23032807537d8dd8ee4ec1c4d40f0633cd4e12f9/packaging/deb-in/extra/mysql-systemd-start#L81-L84
-		mysql_note "Initializing certificates"
-		mysql_ssl_rsa_setup --datadir="$DATADIR"
-		mysql_note "Certificates initialized"
-	fi
 }
 
 # Loads various settings that are used elsewhere in the script
```

**Comment by ltangvald** [comment]
Right. Just need to verify that the resulting datadir is the same. For 5.6 I don't think that block does anything, since it doesn't have this functionality or mysql_ssl_rsa_setup.

**Comment by tianon** [comment]
Updated with the removal of the `mysql_ssl_rsa_setup` code, and verified on 8.0 and 5.7 that certificates are still generated into `/var/lib/mysql`.

8.0:

```console
$ docker exec -it test sh -c 'ls -l /var/lib/mysql/*.pem'
-rw------- 1 mysql mysql 1676 Nov 20 22:21 /var/lib/mysql/ca-key.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:21 /var/lib/mysql/ca.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:21 /var/lib/mysql/client-cert.pem
-rw------- 1 mysql mysql 1680 Nov 20 22:21 /var/lib/mysql/client-key.pem
-rw------- 1 mysql mysql 1680 Nov 20 22:21 /var/lib/mysql/private_key.pem
-rw-r--r-- 1 mysql mysql  452 Nov 20 22:21 /var/lib/mysql/public_key.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:21 /var/lib/mysql/server-cert.pem
-rw------- 1 mysql mysql 1676 Nov 20 22:21 /var/lib/mysql/server-key.pem
```

5.7:

```console
$ docker exec -it test sh -c 'ls -l /var/lib/mysql/*.pem'
-rw------- 1 mysql mysql 1676 Nov 20 22:22 /var/lib/mysql/ca-key.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:22 /var/lib/mysql/ca.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:22 /var/lib/mysql/client-cert.pem
-rw------- 1 mysql mysql 1680 Nov 20 22:22 /var/lib/mysql/client-key.pem
-rw------- 1 mysql mysql 1680 Nov 20 22:22 /var/lib/mysql/private_key.pem
-rw-r--r-- 1 mysql mysql  452 Nov 20 22:22 /var/lib/mysql/public_key.pem
-rw-r--r-- 1 mysql mysql 1112 Nov 20 22:22 /var/lib/mysql/server-cert.pem
-rw------- 1 mysql mysql 1676 Nov 20 22:22 /var/lib/mysql/server-key.pem
```

**Comment by tianon** [comment]
I also verified with `--ssl=0`.

8.0:

```console
$ docker exec -it test sh -c 'ls -l /var/lib/mysql/*.pem'
-rw------- 1 mysql mysql 1680 Nov 20 22:23 /var/lib/mysql/private_key.pem
-rw-r--r-- 1 mysql mysql  452 Nov 20 22:23 /var/lib/mysql/public_key.pem
```

5.7:

```console
$ docker exec -it test sh -c 'ls -l /var/lib/mysql/*.pem'
-rw------- 1 mysql mysql 1680 Nov 20 22:23 /var/lib/mysql/private_key.pem
-rw-r--r-- 1 mysql mysql  452 Nov 20 22:23 /var/lib/mysql/public_key.pem
```

**Review by yosifkit — APPROVED** [review]
This seems fine. Is it safe to merge?

**Pull request merged**